File: Frontend/src/lib/adminWebhookDlq.ts
/**
 * Closed-taxonomy types, Zod schemas and error helpers for the admin
 * **webhook DLQ** surface (`/admin/webhooks/dlq`, T41, ARG-053).
 *
 * Mirrors the canonical pattern established for scan schedules
 * (`adminSchedules.ts`), operations (`adminOperations.ts`) and audit logs
 * (`adminAuditLogs.ts`):
 *
 *   - HTTP status → fixed code (`statusToWebhookDlqActionCode`).
 *   - Backend `detail` token → fixed code (`detailToWebhookDlqActionCode`).
 *     T39 backend uses snake_case identifier tokens (NOT free-text
 *     sentences), so this layer can switch on a closed enum safely.
 *   - Code → user-facing RU sentence (`webhookDlqActionErrorMessage`).
 *   - Dedicated `Error` subclass (`WebhookDlqActionError`) so the React
 *     layer can distinguish action failures from generic JS exceptions
 *     and never surface backend `detail`, stack frames or PII.
 *
 * Wire compatibility (T39 contract, `backend/src/api/routers/admin_webhook_dlq.py`):
 *   - 422 → `validation_failed`
 *   - 409 → `already_replayed` / `already_abandoned`
 *   - 404 → `dlq_entry_not_found` (also for cross-tenant admin probes —
 *           existence-leak protection)
 *   - 403 → `forbidden` / `tenant_required` / `tenant_mismatch`
 *
 * The browser NEVER imports these schemas directly — they live behind
 * `"use server"` actions in `Frontend/src/app/admin/webhooks/dlq/actions.ts`.
 */

import { z } from "zod";

const UUID_RE =
  /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/;

// ---------------------------------------------------------------------------
// Constants — duplicated from `backend/src/api/schemas.py`. Source of truth
// lives on the backend (`WEBHOOK_DLQ_REASON_*` constants); deviations here
// surface as `validation_failed` short-circuits BEFORE the round-trip.
// ---------------------------------------------------------------------------

export const WEBHOOK_DLQ_REASON_MIN = 10;
export const WEBHOOK_DLQ_REASON_MAX = 500;

export const WEBHOOK_DLQ_LIMIT_DEFAULT = 25;
export const WEBHOOK_DLQ_LIMIT_MAX = 200;

export const WEBHOOK_DLQ_TRIAGE_STATUSES = [
  "pending",
  "replayed",
  "abandoned",
] as const;
export type WebhookDlqTriageStatus = (typeof WEBHOOK_DLQ_TRIAGE_STATUSES)[number];

export const WEBHOOK_DLQ_TRIAGE_STATUS_LABELS_RU: Readonly<
  Record<WebhookDlqTriageStatus, string>
> = {
  pending: "В очереди",
  replayed: "Повторено",
  abandoned: "Отброшено",
};

// ---------------------------------------------------------------------------
// Closed taxonomy — UI ONLY ever inspects these codes (plan §7).
// Adding a new code requires a matching entry in `ERROR_MESSAGES_RU` so the
// renderer always finds a human sentence to show.
// ---------------------------------------------------------------------------

export const WEBHOOK_DLQ_FAILURE_TAXONOMY = [
  "unauthorized",
  "forbidden",
  "tenant_required",
  "tenant_mismatch",
  "dlq_entry_not_found",
  "already_replayed",
  "already_abandoned",
  "replay_failed",
  "rate_limited",
  "validation_failed",
  "store_unavailable",
  "server_error",
  "network_error",
] as const;

export const WebhookDlqFailureCodeSchema = z.enum(WEBHOOK_DLQ_FAILURE_TAXONOMY);
export type WebhookDlqFailureCode = z.infer<typeof WebhookDlqFailureCodeSchema>;

export class WebhookDlqActionError extends Error {
  readonly code: WebhookDlqFailureCode;
  readonly status: number | null;

  constructor(code: WebhookDlqFailureCode, status: number | null = null) {
    super(code);
    this.name = "WebhookDlqActionError";
    this.code = code;
    this.status = status;
  }
}

/**
 * Best-effort extraction of a closed-taxonomy code from an unknown error
 * thrown by a server action. Server actions in Next.js cross a
 * serialization boundary that strips the prototype chain — so a thrown
 * `WebhookDlqActionError` arrives on the client as a plain `Error` whose
 * code only survives in `.message` (the `super(code)` constructor call).
 *
 * Resolution order, each step short-circuiting on success:
 *   1. `instanceof WebhookDlqActionError` — direct/in-process callers
 *      (unit tests, server-side composition).
 *   2. `err.code` — defensive, in case future plumbing forwards it.
 *   3. `err.message` — Next.js dev-mode preserves it; in production the
 *      message is stripped but the digest is logged server-side.
 *
 * Returns `null` when no taxonomy member is recognised. Callers should
 * treat `null` as `server_error` for both the displayed message AND any
 * `data-error-code` DOM hook.
 */
export function extractWebhookDlqActionCode(
  err: unknown,
): WebhookDlqFailureCode | null {
  if (err instanceof WebhookDlqActionError) return err.code;
  if (typeof err !== "object" || err === null) return null;
  const candidate = err as { code?: unknown; message?: unknown };
  const taxonomy = WEBHOOK_DLQ_FAILURE_TAXONOMY as readonly string[];
  if (typeof candidate.code === "string" && taxonomy.includes(candidate.code)) {
    return candidate.code as WebhookDlqFailureCode;
  }
  if (typeof candidate.message === "string") {
    const trimmed = candidate.message.trim();
    if (taxonomy.includes(trimmed)) {
      return trimmed as WebhookDlqFailureCode;
    }
  }
  return null;
}

const ERROR_MESSAGES_RU: Readonly<Record<WebhookDlqFailureCode, string>> = {
  unauthorized: "Сессия истекла. Войдите заново.",
  forbidden: "Недостаточно прав для управления DLQ этого tenant.",
  tenant_required:
    "Не указан tenant. Admin обязан выбрать tenant; super-admin — оставить пустым для cross-tenant.",
  tenant_mismatch: "X-Admin-Tenant не совпадает с tenant записи DLQ.",
  dlq_entry_not_found: "Запись DLQ не найдена.",
  already_replayed: "Запись уже была успешно повторена.",
  already_abandoned: "Запись уже была отброшена.",
  replay_failed:
    "Повтор не удался; попытка засчитана. Запись остаётся в DLQ.",
  rate_limited: "Слишком много запросов. Повторите попытку через минуту.",
  validation_failed:
    "Неверные параметры запроса. Проверьте обоснование (10–500 символов).",
  store_unavailable:
    "Backend временно недоступен. Повторите попытку через минуту.",
  server_error: "Не удалось выполнить операцию. Повторите попытку.",
  network_error:
    "Сеть недоступна. Проверьте соединение и повторите попытку.",
};

export function webhookDlqActionErrorMessage(err: unknown): string {
  return ERROR_MESSAGES_RU[extractWebhookDlqActionCode(err) ?? "server_error"];
}

export function getWebhookDlqUserMessage(
  code: WebhookDlqFailureCode | null,
): string {
  return ERROR_MESSAGES_RU[code ?? "server_error"];
}

// ---------------------------------------------------------------------------
// HTTP status / detail-token mapping (plan §7).
// ---------------------------------------------------------------------------

/**
 * HTTP status → closed code. Note that 409 in the DLQ surface is
 * AMBIGUOUS — it can mean `already_replayed` OR `already_abandoned`.
 * We default to `already_replayed` here and override via
 * `detailToWebhookDlqActionCode` when the backend exposes the
 * concrete token.
 */
export function statusToWebhookDlqActionCode(
  status: number,
): WebhookDlqFailureCode {
  if (status === 401) return "unauthorized";
  if (status === 403) return "forbidden";
  if (status === 404) return "dlq_entry_not_found";
  if (status === 409) return "already_replayed";
  if (status === 422) return "validation_failed";
  if (status === 400) return "validation_failed";
  if (status === 429) return "rate_limited";
  // `callAdminBackendJson` collapses transport errors AND missing
  // ADMIN_API_KEY into 503; the user message is identical for both.
  if (status === 503) return "store_unavailable";
  return "server_error";
}

/**
 * Backend `detail` snake_case tokens → closed codes (T39 contract).
 *
 * The backend's `admin_webhook_dlq.py` exports a small fixed set of
 * strings via the `_DETAIL_*` Final constants; we recognise each.
 * Returning `null` means "fall back to the status-based mapping".
 */
export function detailToWebhookDlqActionCode(
  detail: unknown,
): WebhookDlqFailureCode | null {
  if (typeof detail !== "string") return null;
  const normalized = detail.trim().toLowerCase();
  if (normalized === "") return null;

  switch (normalized) {
    case "forbidden":
      return "forbidden";
    case "tenant_required":
    case "tenant_id_required":
    case "tenant_header_required":
      return "tenant_required";
    case "tenant_mismatch":
      return "tenant_mismatch";
    case "dlq_entry_not_found":
      return "dlq_entry_not_found";
    case "already_replayed":
      return "already_replayed";
    case "already_abandoned":
      return "already_abandoned";
    case "replay_failed":
      return "replay_failed";
    case "store_unavailable":
      return "store_unavailable";
    case "server_error":
      return "server_error";
    default:
      return null;
  }
}

/**
 * Top-level mapper used by tests + server actions. Accepts a raw
 * backend response (`{ status, detail }`) and returns a
 * `WebhookDlqActionError` with a closed code.
 */
export function mapWebhookDlqBackendError(input: {
  status: number;
  detail?: unknown;
}): WebhookDlqActionError {
  const fromDetail = detailToWebhookDlqActionCode(input.detail);
  const code = fromDetail ?? statusToWebhookDlqActionCode(input.status);
  return new WebhookDlqActionError(code, input.status);
}

// ---------------------------------------------------------------------------
// Wire schemas (mirror Pydantic models in `backend/src/api/schemas.py`).
// ---------------------------------------------------------------------------

export const WebhookDlqEntryItemSchema = z.object({
  id: z.string().regex(UUID_RE),
  tenant_id: z.string().regex(UUID_RE),
  adapter_name: z.string().min(1),
  event_type: z.string().min(1),
  event_id: z.string().min(1),
  target_url_hash: z.string().min(1),
  attempt_count: z.number().int().min(0),
  last_error_code: z.string(),
  last_status_code: z.number().int().nullable(),
  next_retry_at: z.string().nullable(),
  created_at: z.string(),
  replayed_at: z.string().nullable(),
  abandoned_at: z.string().nullable(),
  abandoned_reason: z.string().nullable(),
  triage_status: z.enum(WEBHOOK_DLQ_TRIAGE_STATUSES),
});

export type WebhookDlqEntryItem = z.infer<typeof WebhookDlqEntryItemSchema>;

export const WebhookDlqListResponseSchema = z.object({
  items: z.array(WebhookDlqEntryItemSchema),
  total: z.number().int().min(0),
  limit: z.number().int().min(1).max(WEBHOOK_DLQ_LIMIT_MAX),
  offset: z.number().int().min(0),
});

export type WebhookDlqListResponse = z.infer<typeof WebhookDlqListResponseSchema>;

export const WebhookDlqReplayResponseSchema = z.object({
  entry_id: z.string().regex(UUID_RE),
  success: z.boolean(),
  attempt_count: z.number().int().min(0),
  new_status: z.enum(["replayed", "pending"]),
  audit_id: z.string().regex(UUID_RE),
  message_code: z.enum(["replay_succeeded", "replay_failed"]),
});

export type WebhookDlqReplayResponse = z.infer<
  typeof WebhookDlqReplayResponseSchema
>;

export const WebhookDlqAbandonResponseSchema = z.object({
  entry_id: z.string().regex(UUID_RE),
  new_status: z.literal("abandoned"),
  audit_id: z.string().regex(UUID_RE),
});

export type WebhookDlqAbandonResponse = z.infer<
  typeof WebhookDlqAbandonResponseSchema
>;

// ---------------------------------------------------------------------------
// Action input schemas
// ---------------------------------------------------------------------------

const ReasonSchema = z
  .string()
  .transform((s) => s.trim())
  .pipe(
    z
      .string()
      .min(WEBHOOK_DLQ_REASON_MIN)
      .max(WEBHOOK_DLQ_REASON_MAX),
  );

export const WebhookDlqReplayInputSchema = z.object({
  reason: ReasonSchema,
});

export type WebhookDlqReplayInput = z.infer<typeof WebhookDlqReplayInputSchema>;

export const WebhookDlqAbandonInputSchema = z.object({
  reason: ReasonSchema,
});

export type WebhookDlqAbandonInput = z.infer<typeof WebhookDlqAbandonInputSchema>;

export type WebhookDlqListFilters = {
  readonly tenantId: string | null;
  readonly status: WebhookDlqTriageStatus | null;
  readonly adapterName: string | null;
  readonly createdAfter: string | null;
  readonly createdBefore: string | null;
  readonly limit: number;
  readonly offset: number;
};

// ---------------------------------------------------------------------------
// UI helpers
// ---------------------------------------------------------------------------

export function isUuid(value: string): boolean {
  return UUID_RE.test(value);
}

/**
 * Truncate a UUID for display: first 8 chars + ellipsis.
 * Mirrors `adminSchedules.shortUuid`.
 */
export function shortUuid(id: string): string {
  if (id.length <= 12) return id;
  return `${id.slice(0, 8)}…`;
}

/**
 * Truncate a SHA-256 fingerprint to its first 12 hex characters. Used
 * for `target_url_hash` rendering — the raw URL is NEVER persisted, so
 * the hash is the only stable identifier the operator can correlate
 * across rows.
 */
export function shortTargetHash(hash: string): string {
  if (hash.length <= 12) return hash;
  return hash.slice(0, 12);
}

File: backend/src/api/routers/admin_webhook_dlq.py
"""T39 (Cycle 6 Batch 5, ARG-053) — webhook DLQ admin API.

Three endpoints:
    GET    /admin/webhooks/dlq                           — list with filters/pagination.
    POST   /admin/webhooks/dlq/{entry_id}/replay         — re-dispatch via NotifierBase.
    POST   /admin/webhooks/dlq/{entry_id}/abandon        — mark as abandoned (operator).

RBAC:
    - operator → 403 forbidden.
    - admin    → own tenant only (X-Admin-Tenant required; entry.tenant_id must match).
    - super-admin → any tenant (X-Admin-Tenant optional; when present, filters list).

Auth envelope: same cookie-shim used by Batches 2-4 (X-Admin-Key + X-Admin-Role
+ X-Admin-Tenant + X-Operator-Subject). ISS-T20-003 (JWT/session-bound auth)
is deferred to Cycle 7 / pre-launch.

Errors map to the closed WEBHOOK_DLQ_FAILURE_TAXONOMY (13 codes). Cross-tenant
probes return 404 dlq_entry_not_found (existence-leak protection) — never a
distinct 403, so admins cannot enumerate other tenants' rows via the
discriminator status code.

Plan: ai_docs/develop/plans/2026-04-22-argus-cycle6-b5.md (T39).
"""

from __future__ import annotations

import logging
from datetime import datetime
from typing import Final, Literal
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import ValidationError

from src.api.routers.admin import require_admin
from src.api.routers.admin_bulk_ops import _operator_subject_dep
from src.api.routers.admin_emergency import _emit_audit
from src.api.routers.admin_findings import _admin_role_dep, _admin_tenant_dep
from src.auth.admin_dependencies import require_admin_mfa_passed
from src.api.schemas import (
    WebhookDlqAbandonRequest,
    WebhookDlqAbandonResponse,
    WebhookDlqEntryItem,
    WebhookDlqListResponse,
    WebhookDlqReplayRequest,
    WebhookDlqReplayResponse,
)
from src.core.observability import tenant_hash, user_id_hash
from src.db.models import WebhookDlqEntry, gen_uuid
from src.db.session import async_session_factory
from src.mcp.services.notifications import (
    AlreadyTerminalError,
    DlqEntryNotFoundError,
    JiraAdapter,
    LinearAdapter,
    NotificationEvent,
    NotifierBase,
    SlackNotifier,
)
from src.mcp.services.notifications import webhook_dlq_persistence as dlq_dao

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Router
# ---------------------------------------------------------------------------

router = APIRouter(
    prefix="/admin/webhooks/dlq",
    tags=["admin", "webhooks-dlq"],
)


# ---------------------------------------------------------------------------
# Closed-taxonomy detail strings (subset of WEBHOOK_DLQ_FAILURE_TAXONOMY).
# Mirrors `Frontend/src/lib/adminWebhookDlq.ts` so the SDK switch on the
# ``detail`` token stays type-safe across the wire. NEVER add a sentence /
# i18n string here — the frontend `detailToWebhookDlqActionCode` table
# expects short snake_case keys only.
# ---------------------------------------------------------------------------

_DETAIL_FORBIDDEN: Final[str] = "forbidden"
_DETAIL_TENANT_REQUIRED: Final[str] = "tenant_required"
_DETAIL_DLQ_ENTRY_NOT_FOUND: Final[str] = "dlq_entry_not_found"
_DETAIL_ALREADY_REPLAYED: Final[str] = "already_replayed"
_DETAIL_ALREADY_ABANDONED: Final[str] = "already_abandoned"
_DETAIL_STORE_UNAVAILABLE: Final[str] = "store_unavailable"
_DETAIL_SERVER_ERROR: Final[str] = "server_error"

#: Audit-log ``action`` taxonomy persisted by these endpoints.
EVENT_DLQ_REPLAY: Final[str] = "webhook_dlq.replay"
EVENT_DLQ_ABANDON: Final[str] = "webhook_dlq.abandon"

#: Closed enum of valid abandon reasons accepted by the DAO. The admin
#: surface always records ``operator`` (free-text justification lives in
#: the audit row, not the persisted ``abandoned_reason`` column).
_ABANDON_REASON_OPERATOR: Final[str] = "operator"

#: Adapter factory — admin replay constructs a fresh adapter per call so
#: it never reuses dispatcher-side circuit-breaker / dedup state. The
#: operator's "force replay" semantically overrides those auto-protections.
_ADAPTER_FACTORY: Final[dict[str, type[NotifierBase]]] = {
    SlackNotifier.name: SlackNotifier,
    LinearAdapter.name: LinearAdapter,
    JiraAdapter.name: JiraAdapter,
}


# ---------------------------------------------------------------------------
# RBAC helpers — local copies that emit closed-taxonomy snake_case detail
# strings (sister routers admin_emergency / admin_schedules use slightly
# different detail vocabularies; keeping ours local pins the contract for
# the WEBHOOK_DLQ_FAILURE_TAXONOMY mirror).
# ---------------------------------------------------------------------------


def _require_admin_or_super(role: str) -> None:
    """Raise 403 ``forbidden`` unless role is ``admin`` or ``super-admin``."""
    if role not in {"admin", "super-admin"}:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=_DETAIL_FORBIDDEN,
        )


def _resolve_effective_tenant(role: str, role_tenant: str | None) -> str | None:
    """Return the tenant scope filter for the DAO call.

    * ``admin``       → ``role_tenant`` MUST be set (else 403 ``tenant_required``).
    * ``super-admin`` → ``role_tenant`` optional; ``None`` means cross-tenant.

    The DAO layer accepts ``tenant_id=None`` as the cross-tenant projection
    and a concrete value as the tenant-scoped projection. Cross-tenant
    probes by an admin (``role_tenant`` does not match the row) return
    ``DlqEntryNotFoundError`` from the DAO so the discriminator status
    code stays at 404 — never a distinct 403 — closing the existence-leak
    side channel.
    """
    if role == "admin":
        if not role_tenant:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=_DETAIL_TENANT_REQUIRED,
            )
        return role_tenant
    return role_tenant


# ---------------------------------------------------------------------------
# Projection helpers
# ---------------------------------------------------------------------------


def _derive_triage_status(
    entry: WebhookDlqEntry,
) -> Literal["pending", "replayed", "abandoned"]:
    """Derive the operator-facing ``triage_status`` from the row state."""
    if entry.replayed_at is not None:
        return "replayed"
    if entry.abandoned_at is not None:
        return "abandoned"
    return "pending"


def _to_entry_item(entry: WebhookDlqEntry) -> WebhookDlqEntryItem:
    """Project an ORM ``WebhookDlqEntry`` onto the response item schema.

    ``target_url_hash`` is forwarded as-is — the raw URL is never
    persisted (see :class:`src.db.models.WebhookDlqEntry`) so the
    frontend can only ever see the opaque fingerprint.
    """
    return WebhookDlqEntryItem(
        id=UUID(entry.id),
        tenant_id=UUID(entry.tenant_id),
        adapter_name=entry.adapter_name,
        event_type=entry.event_type,
        event_id=entry.event_id,
        target_url_hash=entry.target_url_hash,
        attempt_count=entry.attempt_count,
        last_error_code=entry.last_error_code,
        last_status_code=entry.last_status_code,
        next_retry_at=entry.next_retry_at,
        created_at=entry.created_at,
        replayed_at=entry.replayed_at,
        abandoned_at=entry.abandoned_at,
        abandoned_reason=entry.abandoned_reason,
        triage_status=_derive_triage_status(entry),
    )


def _classify_terminal_detail(entry: WebhookDlqEntry) -> str:
    """Return the closed-taxonomy detail token for an already-terminal row."""
    if entry.replayed_at is not None:
        return _DETAIL_ALREADY_REPLAYED
    return _DETAIL_ALREADY_ABANDONED


def _reconstruct_event(entry: WebhookDlqEntry) -> NotificationEvent:
    """Reconstruct the original :class:`NotificationEvent` from ``payload_json``.

    T40 (the daily replay beat task) persists the serialized event
    verbatim into ``payload_json`` so the admin replay path can re-issue
    a byte-for-byte equivalent dispatch. Pydantic validation here is the
    last line of defence against a corrupted row — failures surface as
    500 ``server_error`` (the row data, not the request, is the fault;
    the operator cannot fix it without re-enqueueing or abandoning).
    """
    try:
        return NotificationEvent.model_validate(entry.payload_json)
    except ValidationError as exc:
        logger.error(
            "admin.webhook_dlq.payload_reconstruction_failed",
            extra={
                "event": "argus.admin.webhook_dlq.payload_reconstruction_failed",
                "entry_id": entry.id,
                "adapter_name": entry.adapter_name,
                "tenant_id_hash": tenant_hash(entry.tenant_id),
                "error_count": len(exc.errors()),
            },
        )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=_DETAIL_SERVER_ERROR,
        ) from exc


def _build_adapter(adapter_name: str) -> NotifierBase:
    """Instantiate a fresh adapter for a one-shot operator replay.

    Each replay constructs a dedicated adapter so it never inherits
    long-lived dispatcher state (in-process circuit breaker, dedup
    bucket). Operators clicking "replay" are explicitly bypassing the
    auto-protections that would otherwise short-circuit the call.
    Unknown adapter names are a row-corruption signal and surface as
    500 ``server_error``.
    """
    factory = _ADAPTER_FACTORY.get(adapter_name)
    if factory is None:
        logger.error(
            "admin.webhook_dlq.unknown_adapter",
            extra={
                "event": "argus.admin.webhook_dlq.unknown_adapter",
                "adapter_name": adapter_name,
            },
        )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=_DETAIL_SERVER_ERROR,
        )
    return factory()


# ---------------------------------------------------------------------------
# GET /admin/webhooks/dlq
# ---------------------------------------------------------------------------


@router.get(
    "",
    response_model=WebhookDlqListResponse,
    summary="List webhook DLQ entries (admin own tenant; super-admin cross-tenant)",
)
async def list_webhook_dlq(
    status_filter: Literal["pending", "replayed", "abandoned"] | None = Query(
        default=None,
        alias="status",
        description=(
            "Optional triage-status filter: pending (no terminal column set), "
            "replayed, or abandoned."
        ),
    ),
    adapter_name: str | None = Query(
        default=None, description="Optional exact-match adapter filter (slack/linear/jira)."
    ),
    created_after: datetime | None = Query(
        default=None, description="Lower-bound on ``created_at`` (inclusive)."
    ),
    created_before: datetime | None = Query(
        default=None, description="Upper-bound on ``created_at`` (inclusive)."
    ),
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    _: None = Depends(require_admin),
    role: str = Depends(_admin_role_dep),
    role_tenant: str | None = Depends(_admin_tenant_dep),
    operator_subject: str = Depends(_operator_subject_dep),
) -> WebhookDlqListResponse:
    """Return a paginated DLQ slice for the operator console.

    RBAC: operator → 403; admin → own tenant only (X-Admin-Tenant required);
    super-admin → any tenant (X-Admin-Tenant optional, filters when provided).
    """
    _require_admin_or_super(role)
    effective_tenant = _resolve_effective_tenant(role, role_tenant)

    async with async_session_factory() as session:
        rows, total = await dlq_dao.list_for_tenant(
            session,
            tenant_id=effective_tenant,
            status=status_filter,
            adapter_name=adapter_name,
            created_after=created_after,
            created_before=created_before,
            limit=limit,
            offset=offset,
        )

    items = [_to_entry_item(row) for row in rows]
    logger.info(
        "admin.webhook_dlq.list",
        extra={
            "event": "argus.admin.webhook_dlq.list",
            "user_id_hash": user_id_hash(operator_subject),
            "role": role,
            "tenant_id_hash": (
                tenant_hash(effective_tenant) if effective_tenant else None
            ),
            "status_filter": status_filter,
            "adapter_name": adapter_name,
            "total": total,
            "result_count": len(items),
            "limit": limit,
            "offset": offset,
        },
    )
    return WebhookDlqListResponse(
        items=items, total=total, limit=limit, offset=offset
    )


# ---------------------------------------------------------------------------
# POST /admin/webhooks/dlq/{entry_id}/replay
# ---------------------------------------------------------------------------


@router.post(
    "/{entry_id}/replay",
    response_model=WebhookDlqReplayResponse,
    status_code=status.HTTP_202_ACCEPTED,
    summary="Re-dispatch a DLQ entry via NotifierBase (admin own tenant; super-admin any)",
)
async def replay_webhook_dlq(
    entry_id: UUID,
    body: WebhookDlqReplayRequest,
    _: None = Depends(require_admin_mfa_passed),
    role: str = Depends(_admin_role_dep),
    role_tenant: str | None = Depends(_admin_tenant_dep),
    operator_subject: str = Depends(_operator_subject_dep),
) -> WebhookDlqReplayResponse:
    """Operator-driven replay of a single DLQ entry.

    Steps (single transaction so the audit row + DAO mutation share ACID):

    1. RBAC + tenant scope resolution.
    2. ``get_by_id`` (tenant-scoped) → 404 on miss / cross-tenant.
    3. Already-terminal short-circuit → 409 ``already_replayed`` /
       ``already_abandoned`` (existence-leak safe: cross-tenant rows
       never reach this point).
    4. Reconstruct ``NotificationEvent`` from ``payload_json``;
       corrupted payload → 500 ``server_error``.
    5. Dispatch via a fresh adapter's :meth:`NotifierBase.send_with_retry`
       (intentionally bypasses dispatcher-side circuit / dedup since the
       operator is overriding auto-protections).
    6. On ``delivered=True`` → ``mark_replayed`` (terminal); on failure
       → ``increment_attempt`` (row stays in DLQ).
    7. ``_emit_audit("webhook_dlq.replay", ...)`` and commit.

    The ``adapter.aclose()`` call releases the per-replay httpx client
    eagerly so we do not leak file descriptors under high replay
    concurrency.
    """
    _require_admin_or_super(role)
    effective_tenant = _resolve_effective_tenant(role, role_tenant)
    entry_id_str = str(entry_id)

    async with async_session_factory() as session:
        entry = await dlq_dao.get_by_id(
            session, entry_id=entry_id_str, tenant_id=effective_tenant
        )
        if entry is None:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=_DETAIL_DLQ_ENTRY_NOT_FOUND,
            )
        if entry.replayed_at is not None or entry.abandoned_at is not None:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=_classify_terminal_detail(entry),
            )

        captured_tenant_id = entry.tenant_id
        captured_adapter_name = entry.adapter_name
        captured_event_id = entry.event_id

        event = _reconstruct_event(entry)
        adapter = _build_adapter(captured_adapter_name)
        try:
            result = await adapter.send_with_retry(
                event, tenant_id=captured_tenant_id
            )
        finally:
            await adapter.aclose()

        delivered = bool(result.delivered)
        try:
            if delivered:
                updated = await dlq_dao.mark_replayed(
                    session,
                    entry_id=entry_id_str,
                    tenant_id=effective_tenant,
                )
                new_status: Literal["replayed", "pending"] = "replayed"
                message_code: Literal[
                    "replay_succeeded", "replay_failed"
                ] = "replay_succeeded"
            else:
                updated = await dlq_dao.increment_attempt(
                    session,
                    entry_id=entry_id_str,
                    last_error_code=(
                        result.error_code or result.skipped_reason or "unknown_error"
                    ),
                    last_status_code=result.status_code,
                )
                new_status = "pending"
                message_code = "replay_failed"
        except DlqEntryNotFoundError as exc:
            await session.rollback()
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=_DETAIL_DLQ_ENTRY_NOT_FOUND,
            ) from exc
        except AlreadyTerminalError as exc:
            await session.rollback()
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=_DETAIL_ALREADY_REPLAYED,
            ) from exc

        audit_id = gen_uuid()
        _emit_audit(
            session,
            audit_id=audit_id,
            tenant_id=captured_tenant_id,
            action=EVENT_DLQ_REPLAY,
            operator_subject=operator_subject,
            details={
                "entry_id": entry_id_str,
                "adapter_name": captured_adapter_name,
                "event_id": captured_event_id,
                "tenant_id_hash": tenant_hash(captured_tenant_id),
                "success": delivered,
                "attempt_count": int(updated.attempt_count),
                "reason": body.reason,
                "last_error_code": updated.last_error_code,
                "last_status_code": updated.last_status_code,
            },
        )
        await session.commit()

    logger.info(
        "admin.webhook_dlq.replay",
        extra={
            "event": "argus.admin.webhook_dlq.replay",
            "user_id_hash": user_id_hash(operator_subject),
            "role": role,
            "tenant_id_hash": tenant_hash(captured_tenant_id),
            "entry_id": entry_id_str,
            "adapter_name": captured_adapter_name,
            "success": delivered,
            "attempt_count": int(updated.attempt_count),
            "audit_id": audit_id,
            "message_code": message_code,
        },
    )
    return WebhookDlqReplayResponse(
        entry_id=entry_id,
        success=delivered,
        attempt_count=int(updated.attempt_count),
        new_status=new_status,
        audit_id=UUID(audit_id),
        message_code=message_code,
    )


# ---------------------------------------------------------------------------
# POST /admin/webhooks/dlq/{entry_id}/abandon
# ---------------------------------------------------------------------------


@router.post(
    "/{entry_id}/abandon",
    response_model=WebhookDlqAbandonResponse,
    status_code=status.HTTP_200_OK,
    summary="Mark a DLQ entry abandoned (admin own tenant; super-admin any)",
)
async def abandon_webhook_dlq(
    entry_id: UUID,
    body: WebhookDlqAbandonRequest,
    _: None = Depends(require_admin_mfa_passed),
    role: str = Depends(_admin_role_dep),
    role_tenant: str | None = Depends(_admin_tenant_dep),
    operator_subject: str = Depends(_operator_subject_dep),
) -> WebhookDlqAbandonResponse:
    """Operator-driven abandon of a single DLQ entry.

    Steps (single transaction so the audit row + DAO mutation share ACID):

    1. RBAC + tenant scope resolution.
    2. ``get_by_id`` (tenant-scoped) → 404 on miss / cross-tenant.
    3. Already-terminal short-circuit → 409 ``already_replayed`` /
       ``already_abandoned``.
    4. ``mark_abandoned(reason='operator')`` flips ``abandoned_at`` +
       ``abandoned_reason`` (terminal). Free-text justification from
       ``body.reason`` is recorded ONLY in the audit details (not in
       the persisted ``abandoned_reason`` enum) — operators sometimes
       paste customer-sensitive context here, and the audit row is the
       proper home for it.
    5. ``_emit_audit("webhook_dlq.abandon", ...)`` and commit.
    """
    _require_admin_or_super(role)
    effective_tenant = _resolve_effective_tenant(role, role_tenant)
    entry_id_str = str(entry_id)

    async with async_session_factory() as session:
        entry = await dlq_dao.get_by_id(
            session, entry_id=entry_id_str, tenant_id=effective_tenant
        )
        if entry is None:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=_DETAIL_DLQ_ENTRY_NOT_FOUND,
            )
        if entry.replayed_at is not None or entry.abandoned_at is not None:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=_classify_terminal_detail(entry),
            )

        captured_tenant_id = entry.tenant_id
        captured_adapter_name = entry.adapter_name
        captured_event_id = entry.event_id

        try:
            await dlq_dao.mark_abandoned(
                session,
                entry_id=entry_id_str,
                tenant_id=effective_tenant,
                reason=_ABANDON_REASON_OPERATOR,
            )
        except DlqEntryNotFoundError as exc:
            await session.rollback()
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=_DETAIL_DLQ_ENTRY_NOT_FOUND,
            ) from exc
        except AlreadyTerminalError as exc:
            await session.rollback()
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=_DETAIL_ALREADY_ABANDONED,
            ) from exc

        audit_id = gen_uuid()
        _emit_audit(
            session,
            audit_id=audit_id,
            tenant_id=captured_tenant_id,
            action=EVENT_DLQ_ABANDON,
            operator_subject=operator_subject,
            details={
                "entry_id": entry_id_str,
                "adapter_name": captured_adapter_name,
                "event_id": captured_event_id,
                "tenant_id_hash": tenant_hash(captured_tenant_id),
                "reason": body.reason,
                "abandoned_reason": _ABANDON_REASON_OPERATOR,
            },
        )
        await session.commit()

    logger.info(
        "admin.webhook_dlq.abandon",
        extra={
            "event": "argus.admin.webhook_dlq.abandon",
            "user_id_hash": user_id_hash(operator_subject),
            "role": role,
            "tenant_id_hash": tenant_hash(captured_tenant_id),
            "entry_id": entry_id_str,
            "adapter_name": captured_adapter_name,
            "audit_id": audit_id,
        },
    )
    return WebhookDlqAbandonResponse(
        entry_id=entry_id,
        new_status="abandoned",
        audit_id=UUID(audit_id),
    )


__all__ = [
    "EVENT_DLQ_ABANDON",
    "EVENT_DLQ_REPLAY",
    "abandon_webhook_dlq",
    "list_webhook_dlq",
    "replay_webhook_dlq",
    "router",
]
